test/python/c_api/test_transpile.py

```diff
@@ -0,0 +1,304 @@
+        self.assertGreaterEqual(set(basis) | {"barrier"}, transpiled.count_ops().keys())
+        self.assertEqual(Operator(qc), Operator(transpiled))
+
+    @ddt.data(0, 1, 2, 3)
+    def test_single_qubit_circuit_deterministic_output(self, optimization_level):
+        """Test that the transpiler's output is deterministic in a single qubit example.
+
+        Reproduce from `#14729 <https://github.com/Qiskit/qiskit/issues/14729>`__"""
+        params = [math.pi, math.pi / 2, math.pi * 2, math.pi / 4, 0]
+
+        circ = QuantumCircuit(len(params))
+        for i, par in enumerate(params):
+            circ.rx(par, i)
+        circ.measure_all()
+        target = ffi.build_homogenous_target(
+            CouplingMap.from_full(10), ["cx", "rz", "sx", "x"], 123
+        )
+        c_circ = ffi.build_circuit_from_python(circ)
+        isa_circs = []
+        for _ in range(10):
+            isa_circs.append(ffi.transpile_from_c(c_circ, target, optimization_level, 1.0, 123))
+        for i in range(10):
+            self.assertEqual(isa_circs[0], isa_circs[i])
+
+    @ddt.data(2, 3)
+    def test_size_optimization(self, level):
+        """Test the levels for optimization based on size of circuit"""
+        target = ffi.build_homogenous_target(CouplingMap.from_full(8), ["u3", "cx"], 42, True)
+        qc = QuantumCircuit(8)
+        qc.cx(1, 2)
+        qc.cx(2, 3)
+        qc.cx(5, 4)
+        qc.cx(6, 5)
+        qc.cx(4, 5)
+        qc.cx(3, 4)
+        qc.cx(5, 6)
+        qc.cx(5, 4)
+        qc.cx(3, 4)
+        qc.cx(2, 3)
+        qc.cx(1, 2)
+        qc.cx(6, 7)
+        qc.cx(6, 5)
+        qc.cx(5, 4)
+        qc.cx(7, 6)
+        qc.cx(6, 7)
+        c_circ = ffi.build_circuit_from_python(qc)
+        circ = ffi.transpile_from_c(c_circ, target, level, 1.0, 123)
+
+        circ_data = circ.data
+        free_qubits = {0, 1, 2, 3}
+
+        # ensure no gates are using qubits - [0,1,2,3]
+        for gate in circ_data:
+            layout = circ.layout.initial_layout
+            indices = {layout[circ.find_bit(qubit).index] for qubit in gate.qubits}
+            common = indices.intersection(free_qubits)
+            for common_qubit in common:
+                self.assertTrue(common_qubit not in free_qubits)
+
+        self.assertLess(circ.size(), qc.size())
+        self.assertLessEqual(circ.depth(), qc.depth())
+
+    @ddt.data(0, 1, 2, 3)
+    def test_1(self, level):
+        """Simple coupling map (linear 5 qubits)."""
+        basis = ["u1", "u2", "cx", "swap"]
+        coupling_map = CouplingMap([(0, 1), (1, 2), (2, 3), (3, 4)])
+        circuit = QuantumCircuit(5)
+        circuit.cx(2, 4)
+        c_circ = ffi.build_circuit_from_python(circuit)
+        target = ffi.build_homogenous_target(coupling_map, basis, seed=24)
+        result = ffi.transpile_from_c(c_circ, target, level, 1.0, 123)
+
+        self.assertIsInstance(result, QuantumCircuit)
+        self.assertEqual(Operator.from_circuit(result), Operator(circuit))
```

Comment:
this triggers dev OCD (same below) 😛 Since above we ensure the type is `QuantumCircuit` just this works too, no?
```suggestion
        self.assertEqual(Operator(result), Operator(circuit))
```

Reply:
`from_circuit` is layout aware, that's why I used it here. You can pass a circuit to the operator constructor but it doesn't know how to work with the layout

Reply:
Wait there's a difference between `Operator(circuit)` and `Operator.from_circuit(circuit)`? It feels like the former should just forward to the latter, no?

Reply:
It does not, this was for backwards compatibility. `Operator(circuit)` is not layout aware. We didn't forward it because if someone was using it before the output operator was not layout aware, if they upgraded the version the operator they would get would be completely different. This is documented iirc in the operator class and the `from_circuit` method.

Reply:
🥲